### PR TITLE
Track git repository URL when creating MLflow Runs from MLflow Pipelines

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Track git repository URL when creating MLflow Runs from MLflow Pipelines

## How is this patch tested?

Unit tests, manual verification:

Screen Shot 2022-06-27 at 1 33 40 PM

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

When running an MLflow Pipeline, the URL of the repository is now tracked with the corresponding MLflow Run.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [X] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/git_utils.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    from git import Repo
+
+    try:
+        repo = Repo(path, search_parent_directories=True)
+        remote_urls = [remote.url for remote in repo.remotes]
+        if len(remote_urls) == 0:
+            return None
+    except Exception:
+        return None
+    return remote_urls[0]
```
Comment (harupy) [inline mlflow/utils/git_utils.py:18]:
```suggestion
    try:
        repo = Repo(path, search_parent_directories=True)
        return next(remote.url for remote in repo.remotes, None)
    except Exception:
        return None
```

Can we use `next(..., None)` to simplify the code?

**Review by harupy — APPROVED** [review]
Left a minor comment, otherwise LGTM!

File: mlflow/tracking/context/git_context.py [hunk 2]
```diff
@@ -45,8 +57,18 @@ def _source_version(self):
...
+        return self._cache["git_repo_url"]
+
     def in_context(self):
         return self._source_version is not None
 
     def tags(self):
-        return {MLFLOW_GIT_COMMIT: self._source_version}
+        return {
+            MLFLOW_GIT_COMMIT: self._source_version,
+            MLFLOW_GIT_REPO_URL: self._git_repo_url,
```
Comment (jinzhang21) [inline mlflow/tracking/context/git_context.py:72]: Do you plan to add `MLFLOW_GIT_BRANCH` in a follow up PR?

**Review by jinzhang21 — APPROVED** [review]

File: mlflow/tracking/context/git_context.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/tracking/context/git_context.py]: Added it to this one! Thanks :)

**Pull request merged**